lib/cfp.js

```diff
@@ -0,0 +1,325 @@
+        ...{
+          [data.name]: {
+            value: data.value,
+            doc,
+          },
+        },
+      };
+    },
+    {},
+  );
+
+  options[OPTIONS.SEARCH_HASHTAGS] = {
+    value: extractHashtags(options[OPTIONS.SEARCH_VALUE].value),
+    doc: null,
+  };
+
+  return options;
+};
+
+const updateLastResultId = async (id, db = null) => {
+  const useDB = db || getDB();
+  const collection = useDB.collection(COLLECTIONS.OPTIONS);
+  const savedOption = await collection.where('name', '==', OPTIONS.LAST_RESULT_ID).get();
+  if (savedOption.empty) {
+    await collection.add({
+      name: OPTIONS.LAST_RESULT_ID,
+      value: id,
+    });
+  } else {
+    const [doc] = savedOption.docs;
+    await collection.doc(doc.id).update({ value: id });
+  }
+};
+
+const extractTweetHashtags = async (tweet, options = null) => {
+  if (tweet.entities && tweet.entities.hashtags) {
+    let useOptions;
+    if (options) {
+      useOptions = options;
+    } else {
+      useOptions = await getOptions();
+    }
+
+    return tweet.entities.hashtags
+      .map(({ text }) => text.toLowerCase())
+      .filter((hashtag) => !useOptions[OPTIONS.SEARCH_HASHTAGS].value.includes(hashtag))
+      .reduce((acc, hashtag) => (acc.includes(hashtag) ? acc : [...acc, hashtag]), []);
+  }
+
+  return [];
+};
+
+const formatTweet = async (tweet, options = null) => ({
+  tweet: {
+    tweetId: tweet.id_str,
+    username: tweet.user.screen_name,
+    tweet: tweet.text,
+    url: `https://twitter.com/${tweet.user.screen_name}/status/${tweet.id_str}`,
+    time: new Date(),
+  },
+  hashtags: await extractTweetHashtags(tweet, options),
+  doc: null,
+  isNew: true,
+});
+
+const formatTweets = async (tweets, options = null) => Promise.all(tweets.map(async (tweet) => {
+  const newTweet = await formatTweet(tweet, options);
+  return newTweet;
+}));
+
+const getBlacklist = async (db = null) => {
+  const useDB = db || getDB();
+  const results = await useDB.collection(COLLECTIONS.BLACKLIST).get();
+  return results.docs.reduce(
+    (acc, doc) => {
+      const nextAcc = { ...acc };
+      const data = doc.data();
+      if (data.hashtag) {
+        nextAcc.hashtags.push(data.hashtag);
+      }
+      if (data.username) {
+        nextAcc.usernames.push(data.username);
+      }
+
+      return nextAcc;
+    },
+    {
+      hashtags: [],
+      usernames: [],
+    },
+  );
+};
+
+const filterBlacklistedTweets = async (formattedTweets, db = null) => {
+  const blacklist = await getBlacklist(db);
+  return formattedTweets.filter((info) => {
+    if (blacklist.usernames.includes(info.tweet.username)) {
+      return false;
+    }
+
+    if (info.hashtags.some((hashtag) => blacklist.hashtags.includes(hashtag))) {
+      return false;
+    }
+
+    return true;
+  });
+};
+
+const saveTweet = async (formattedTweet, db = null) => {
+  const useDB = db || getDB();
+  const collection = useDB.collection(COLLECTIONS.TWEETS);
+  let doc;
+  let isNew = false;
```

Comment:
This flag seems a bit superfluous and could be mapped directly to the value of `savedTweet.empty`:

```js
const isNew = savedTweet.empty;
if (isNew) {
  // ...
} else {
  // ...
}
```